feat(files_trashbin): Refactor expire background job to support parallel run

artonge opened the pull request:
- Follow-up of https://github.com/nextcloud/server/pull/51600

The original PR introduced the possibility to continue an `ExpireTrash` job by saving the offset. This was to prevent having to start over the whole user list when the job crashed or was killed. 

But on big instances, one process is not enough to go through all the users in a timely manner. Supporting parallel run allows covering more ground faster.

This PR introduced this possibility. We are now storing the offset right away to allow another parallel job to pick up the task at that point. We are arbitrarily cutting the user list in chunk of 10 to not drastically overflow the 30 minutes time limit.

come-nc commented:
`apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:49:4: AssignmentToVoid: Cannot assign $offset to type void (see https://psalm.dev/121)
`
Sounds real

sgyuris commented:
1. The 1. job starts, query for 10 user and write down the offset, let's say: 10
2. The 2. job starts and query for 10 user with offset of 10. Immediately it writes down the offset: 20
3. The 1. job terminates just after it started to clean up user3's trashbin.

What will happen with the remaining files in the user3's trashbin?
What will happen with the user4 - user10 trashbins?

artonge commented:
> 3\. The 1. job terminates just after it started to clean up user3's trashbin.

How would that happen ?

sgyuris commented:
> > 3. The 1. job terminates just after it started to clean up user3's trashbin.
> 
> How would that happen ?

MySQL has gone away... or db deadlock... storage issues or it is killed by any other means like oomkiller...

I know this indicates some kind of infrastructure error still we need to account for it.

artonge commented:
> MySQL has gone away... or db deadlock... storage issues or it is killed by any other means like oomkiller...

All of those should be caught by the `try/catch`.

sgyuris commented:
Lets see...

artonge commented:
/backport to stable31

artonge commented:
/backport to stable30

artonge commented:
/backport! to stable29

artonge commented:
/backport! to stable30

artonge commented:
/backport! to stable31

artonge committed:
feat: Support limit argument in `getSeenUsers`

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
/backport! to stable29

artonge commented:
/backport! to stable30

artonge commented:
/backport! to stable31

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,44 +45,47 @@ protected function run($argument) {
+		do {
+			$this->appConfig->clearCache();
+			$offset = $this->appConfig->getValueInt('files_trashbin', 'background_job_expire_trash_offset', 0);
+			$this->appConfig->setValueInt('files_trashbin', 'background_job_expire_trash_offset', $offset + 10);
+
+			$users = $this->userManager->getSeenUsers($offset, 10);
```
The limit should be a variable/constant.

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,44 +45,47 @@ protected function run($argument) {
+
+		do {
+			$this->appConfig->clearCache();
+			$offset = $this->appConfig->getValueInt('files_trashbin', 'background_job_expire_trash_offset', 0);
+			$this->appConfig->setValueInt('files_trashbin', 'background_job_expire_trash_offset', $offset + 10);
```
Since there is no lock, this could go wrong when multiple parallel jobs are reading/writing, no?

artonge replied:
Ended up using an already available locking solution

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,44 +45,47 @@ protected function run($argument) {
 			}
 
-			$offset++;
+		} while (time() < $stopTime && $count === 10);
```
I guess this 10 is the same 10 as above? Then it should also use the same variable.

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,44 +45,47 @@ protected function run($argument) {
-				\OC_Util::tearDownFS();
-				return;
-			}
+		if ($count < 10) {
```
And here again

artonge committed:
feat(files_trashbin): Refactor expire background job to support parallel run

- Follow-up of #51600

The original PR introduced the possibility to continue an `ExpireTrash` job by saving the offset. This was to prevent having to start over the whole user list when the job crashed or was killed.

But on big instances, one process is not enough to go through all the users in a timely manner. Supporting parallel run allows covering more ground faster.

This PR introduced this possibility. We are now storing the offset right away to allow another parallel job to pick up the task at that point. We are arbitrarily cutting the user list in chunk of 10 to not drastically overflow the 30 minutes time limit.

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,48 +48,86 @@ protected function run($argument) {
 			}
 
-			$offset++;
+		} while (time() < $stopTime && $count === self::USER_BATCH_SIZE);
```
Fair, it was a bit obscure. Rewrote in an attempt to make more explicit. Let me know if it helps.

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,48 +48,89 @@ protected function run($argument) {
+
+		while ($acquired === false) {
+			try {
+				$this->lockingProvider->acquireLock('background_job_expire_trash', ILockingProvider::LOCK_EXCLUSIVE, 'Expire trashbin background job');
+				$acquired = true;
+			} catch (\OCP\Lock\LockedException $e) {
+				// wait a bit and try again
+				usleep(100000);
+			}
+		}
+
+		try {
+			$result = $operation();
+		} finally {
+			$this->lockingProvider->releaseLock('background_job_expire_trash', ILockingProvider::LOCK_EXCLUSIVE);
```
```suggestion
				$this->lockingProvider->acquireLock('background_job_expire_trash_offset', ILockingProvider::LOCK_EXCLUSIVE, 'Expire trashbin background job offset');
				$acquired = true;
			} catch (\OCP\Lock\LockedException $e) {
				// wait a bit and try again
				usleep(100000);
			}
		}

		try {
			$result = $operation();
		} finally {
			$this->lockingProvider->releaseLock('background_job_expire_trash_offset', ILockingProvider::LOCK_EXCLUSIVE);
```
Just to make it a bit clearere that it's about the offset.

provokateurin replied:
And then also move the keys to constants.

provokateurin commented on apps/files_trashbin/lib/BackgroundJob/ExpireTrash.php:
```diff
@@ -41,48 +48,89 @@ protected function run($argument) {
+		return $this->runMutexOperation(function () {
+			$this->appConfig->clearCache();
+
+			$offset = $this->appConfig->getValueInt('files_trashbin', 'background_job_expire_trash_offset', 0);
+			$this->appConfig->setValueInt('files_trashbin', 'background_job_expire_trash_offset', $offset + self::USER_BATCH_SIZE);
+
+			return $offset;
+		});
+
+	}
+
+	private function resetOffset() {
+		$this->runMutexOperation(function () {
+			$this->appConfig->setValueInt('files_trashbin', 'background_job_expire_trash_offset', 0);
```
Please move the keys to constants and use \OCA\Files_Trashbin\AppInfo\Application::APP_ID

artonge commented:
/backport to stable32

artonge commented:
/backport to stable31
